Index: render_view_impl.cc
===================================================================
--- render_view_impl.cc	(revision 121456)
+++ render_view_impl.cc	(working copy)
@@ -2204,7 +2204,10 @@
   // script, drag-n-drop operation, etc.
   bool is_content_initiated =
       DocumentState::FromDataSource(frame->provisionalDataSource())->
-          navigation_state()->is_content_initiated();
+          navigation_state()->is_content_initiated() &&
+      // May be set to a non-zero value by embedders originating requests in
+      // the renderer process.
+      request.requestorID() == 0;
 
   // Experimental:
   // If --enable-strict-site-isolation is enabled, send all top-level
